ocaml/xapi/api_server.ml

```diff
@@ -250,24 +250,27 @@ let is_host_is_slave_error (response : Rpc.response) =
       false
 
 let create_thumbprint_header req response =
-  let include_thumbprint =
+  let hash_type_option =
     match
       List.assoc_opt
         !Xapi_globs.cert_thumbprint_header_request
         req.Http.Request.additional_headers
     with
-    | Some x when x = !Xapi_globs.cert_thumbprint_header_value ->
-        true
+    | Some x when x = !Xapi_globs.cert_thumbprint_header_value_sha256 ->
+        Some `Sha256
+    | Some x when x = !Xapi_globs.cert_thumbprint_header_value_sha1 ->
+        Some `Sha1
     | _ ->
-        false
+        None
   in
-  if include_thumbprint && is_host_is_slave_error response then
-    Helpers.external_certificate_thumbprint_of_master ()
-    |> Option.fold ~none:[] ~some:(fun x ->
-           [(!Xapi_globs.cert_thumbprint_header_response, x)]
-       )
-  else
-    []
+  match (hash_type_option, is_host_is_slave_error response) with
```

Comment:
this is an interesting line: we want to enforce an order of operations here, to prevent unnecessary work as this is executed at every response:
1. check if the header is in the request (assumed to be very cheap to check)
2. if it is, then check if the response error is host_is_slave
3. if it is, then fetch the certificate thumbprint
4. if there is one, then create the response header, otherwise just return []

this order is kind of visible in the original code, less so after this patch. In order to preserve this order in a more explicit way, I wonder if it would be possible to use a more explicit ordering of these operations, here, something like
```
some req
Option.bind hash_type (** req -> [Sha1|Sha256] option **)
Option.bind with_host_is_slave_error response   (** response -> `a  -> `a option **) 
Option.bind with_external_certificate_thumbprint_of_master (** hash_type -> thumbprint option **)
Option.fold
    ~some:(fun x-> [(!Xapi_globs.cert_thumbprint_header_response, x)] )
    ~none:[] 
```